core/src/main/java/com/linecorp/armeria/common/stream/ConcatPublisherStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+            final StreamMessage<? extends T> streamMessage = StreamMessage.of(publisher);
+            currentPublisher = streamMessage;
+            streamMessage.subscribe(innerSubscriber, executor, innerSubscriber.options);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (completed) {
+                return;
+            }
+            completed = true;
+            innerSubscriber.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (completed) {
+                return;
+            }
+
+            // If 'innerSubscriber' is not complete, 'downstream' will be completed
+            // when 'innerSubscriber' receives `onComplete` signal.
+            completed = true;
+            if (!inSubscribe && innerSubscriber.completed) {
+                innerSubscriber.onComplete();
+            }
+        }
+
+        void nextSource() {
+            upstream.request(1);
+        }
+    }
+
+    private static final class InnerSubscriber<T> extends SubscriptionArbiter implements Subscriber<T> {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<InnerSubscriber> cancelledUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(InnerSubscriber.class, "cancelled");
+
+        private Subscriber<? super T> downstream;
+        private final SubscriptionOption[] options;
+
+        @Nullable
+        private OuterSubscriber<T> outerSubscriber;
+
+        private volatile boolean completed = true;
+        // Updated via cancelledUpdater
+        private volatile int cancelled;
+
+        InnerSubscriber(Subscriber<? super T> downstream, SubscriptionOption[] options) {
+            this.downstream = downstream;
+            this.options = options;
+        }
+
+        /**
+         * Sets the {@link OuterSubscriber}.
+         * This should be set before subscribing to upstream.
+         */
+        void setOuterSubscriber(OuterSubscriber<T> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (isCancelled()) {
+                subscription.cancel();
+                return;
+            }
+            // Reset 'completed' to subscribe to new Publisher
+            completed = false;
+            setUpstreamSubscription(subscription);
+            outerSubscriber.inSubscribe = false;
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            if (isCancelled()) {
+                return;
+            }
+            downstream.onNext(item);
+            produced(1);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            downstream.onError(throwable);
```

Comment:
Shouldn't we cancel all other unsubcribed stream message?

Reply:
That is a good point. I've added some code aborting publishers to `RegularStreamMessage`, `DefaultStreamMessage` and `PublisherBasedStreamMesage`.